extensions/amp-viewer-integration/0.1/test/integration/test-messaging-handshake.js

```diff
@@ -130,5 +108,25 @@ describes.sandboxed('AmpViewerMessagingIntegration', {}, () => {
             expect(name).to.equal('documentLoaded');
           });
       });
+
+      it('should fail if messaging token is wrong', () => {
+        const params = serializeQueryString({
+          origin: getWinOrigin(window),
+          messagingToken: 'foo',
+        });
+        const ampDocUrl = `${iframeOrigin}${ampDocSrc}#${params}`;
+        viewerIframe.setAttribute('src', ampDocUrl);
+
+        return Messaging.waitForHandshakeFromDocument(
+          window,
+          viewerIframe.contentWindow,
+          iframeOrigin,
+          'bar'
+        ).then((messaging) => {
+          const handlerStub = window.sandbox.stub();
+          messaging.setDefaultHandler(handlerStub);
+          expect(handlerStub).to.not.have.been.called;
```

Comment:
@zhouyx Your observation about the test not finishing its promises makes me think we should be returning the result of `messaging.setDefaultHandler(handlerStub);` in the `then` block and do the `expect` in the following block, in order to guarantee that the result is available before executing the `expect`. Alternatively, for cleaner code, you could make the test function `async` and add a bunch of `await`s in the function body.

Reply:
The `setDefaultHandler` is a synchronous method. I tried using await, it didn't help. 

What would be the difference that's causing the test to fail only in headless browser.